### ISSUE TITLE
Perl's stdio layer doesn't handle EINTR

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6248]
Migrated from [rt.perl.org#20525](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=20525) (status was 'resolved')

Searchable as RT20525$

**Comment by p5pRT** [comment]
### From mls@suse.de

#### Created by mls@suse.de



Perl's stdio IO layer doesn't restart the library calls if
they return EINTR\. Old perl versions aren't affected as they
use SA\_RESTART when they set up the signal handlers\.

Test program&#8203;:

use POSIX;

$SIG\{CHLD\} = sub \{\};

if \(fork\(\) == 0\) \{
&nbsp;  $ppid = getppid;
&nbsp;  while \(1\) \{
&nbsp;    exit if kill\(POSIX&#8203;::SIGCHLD\, $ppid\) \!= 1;
&nbsp;  \}
\}
$cnt = 0;
open\(F\, "cat /etc/termcap|"\) || die "open\\n";
$cnt\+\+ while \<F>;
close F;
print "\-> $cnt\\n";

The test program often returns zero read lines \(GNU glibc\)\. Adding
&nbsp;    binmode F\, "&#8203;:unix"
fixes this\, of course\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

This perlbug was built using Perl v5.8.0 - Mon Sep  9 18:12:37 UTC 2002
It is being executed now by  Perl v5.8.0 - Mon Sep  9 18:02:36 UTC 2002.

Site configuration information for perl v5.8.0:

Configured by root at Mon Sep  9 18:02:36 UTC 2002.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.19, archname=i586-linux-thread-multi
    uname='linux bloembergen 2.4.19 #1 mon apr 15 08:57:26 gmt 2002 i686 unknown '
    config_args='-ds -e -Dprefix=/usr -Dusethreads -Di_db -Di_dbm -Di_ndbm -Di_gdbm -Duseshrplib=true'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O3 --pipe',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing'
    ccversion='', gccversion='3.2', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =''
    libpth=/lib /usr/lib /usr/local/lib
    libs=-lnsl -ldl -lm -lpthread -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lpthread -lc -lcrypt -lutil
    libc=, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.2.5'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic -Wl,-rpath,/usr/lib/perl5/5.8.0/i586-linux-thread-multi/CORE'
    cccdlflags='-fPIC', lddlflags='-shared'

Locally applied patches:
    


@INC for perl v5.8.0:
    /usr/lib/perl5/5.8.0/i586-linux-thread-multi
    /usr/lib/perl5/5.8.0
    /usr/lib/perl5/site_perl/5.8.0/i586-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.0
    /usr/lib/perl5/site_perl
    .


Environment for perl v5.8.0:
    HOME=/suse/mls
    LANG (unset)
    LANGUAGE (unset)
    LC_COLLATE=POSIX
    LC_CTYPE=de_DE@euro
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/suse/mls/bin:/opt/kde3/bin:/usr/local/bin:/usr/bin:/usr/sbin:/sbin:/usr/X11R6/bin:/bin:/usr/lib/java/bin:/usr/games/bin:/usr/games:/opt/gnome/bin:/opt/kde/bin:/usr/openwin/bin:/opt/pilotsdk/bin:/suse/mls/korn
    PERL_BADLANG (unset)
    SHELL=/bin/tcsh

```
</details>

**Comment by p5pRT** [comment]
### From nick@ing-simmons.net

Mls@&#8203;Suse\.De \<perl5\-porters@&#8203;perl\.org> writes&#8203;:
>\# New Ticket Created by  mls@&#8203;suse\.de 
>\# Please include the string&#8203;:  \[perl \#20525\]
>\# in the subject line of all future correspondence about this issue\. 
>\# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt2/Ticket/Display\.html?id=20525 >
>
>
>
>This is a bug report for perl from mls@&#8203;suse\.de\,
>generated with the help of perlbug 1\.34 running under perl v5\.8\.0\.
>
>
>\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>\[Please enter your report here\]
>
>Perl's stdio IO layer doesn't restart the library calls if
>they return EINTR\. Old perl versions aren't affected as they
>use SA\_RESTART when they set up the signal handlers\.

This is a side effect of the "safe signals" support\.
I "had" to turn off SA\_RESTART so that syscall would resturn and give us 
a chance to execute the perl level %SIG handler\.
:perlio layer attempts to re\-do \(as opposed to re\-start\) the system call\,
but I have not \(yet?\) tried that for :stdio as it is harder to 2nd\-guess
the system's stdio as to where it has got to\.

So using :perlio is a possible workround in short term\.

A possible alternative to the above would be to have a global flag 
which gets set just before we start a syscall to say calling %SIG handler
is okay \(as we are in a syscall not malloc or perl code\) and re\-instate
SA\_RESTART\. The snag is that un\-setting the flag is a potential race\.
must cope with die and needs splattering all over the C code\.
Potentially we can use PerlSIO\_xxx/PerlLIO\_xxx wrappers to do this\.



>
>Test program&#8203;:
>
>use POSIX;
>
>$SIG\{CHLD\} = sub \{\};
>
>if \(fork\(\) == 0\) \{
>  $ppid = getppid;
>  while \(1\) \{
>    exit if kill\(POSIX&#8203;::SIGCHLD\, $ppid\) \!= 1;
>  \}
>\}
>$cnt = 0;
>open\(F\, "cat /etc/termcap|"\) || die "open\\n";
>$cnt\+\+ while \<F>;
>close F;
>print "\-> $cnt\\n";
>
>The test program often returns zero read lines \(GNU glibc\)\. Adding
>    binmode F\, "&#8203;:unix"
>fixes this\, of course\.
>
>\[Please do not change anything below this line\]
>\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>\-\-\-
>Flags&#8203;:
>    category=core
>    severity=medium
>\-\-\-
>This perlbug was built using Perl v5\.8\.0 \- Mon Sep  9 18&#8203;:12&#8203;:37 UTC 2002
>It is being executed now by  Perl v5\.8\.0 \- Mon Sep  9 18&#8203;:02&#8203;:36 UTC 2002\.
>
>Site configuration information for perl v5\.8\.0&#8203;:
>
>Configured by root at Mon Sep  9 18&#8203;:02&#8203;:36 UTC 2002\.
>
>Summary of my perl5 \(revision 5\.0 version 8 subversion 0\) configuration&#8203;:
>  Platform&#8203;:
>    osname=linux\, osvers=2\.4\.19\, archname=i586\-linux\-thread\-multi
>    uname='linux bloembergen 2\.4\.19 \#1 mon apr 15 08&#8203;:57&#8203;:26 gmt 2002 i686 unknown '
>    config\_args='\-ds \-e \-Dprefix=/usr \-Dusethreads \-Di\_db \-Di\_dbm \-Di\_ndbm \-Di\_gdbm \-Duseshrplib=true'
>    hint=recommended\, useposix=true\, d\_sigaction=define
>    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
>    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
>    use64bitint=undef use64bitall=undef uselongdouble=undef
>    usemymalloc=n\, bincompat5005=undef
>  Compiler&#8203;:
>    cc='cc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-fno\-strict\-aliasing \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
>    optimize='\-O3 \-\-pipe'\,
>    cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-fno\-strict\-aliasing'
>    ccversion=''\, gccversion='3\.2'\, gccosandvers=''
>    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
>    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
>    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
>    alignbytes=4\, prototype=define
>  Linker and Libraries&#8203;:
>    ld='cc'\, ldflags =''
>    libpth=/lib /usr/lib /usr/local/lib
>    libs=\-lnsl \-ldl \-lm \-lpthread \-lc \-lcrypt \-lutil
>    perllibs=\-lnsl \-ldl \-lm \-lpthread \-lc \-lcrypt \-lutil
>    libc=\, so=so\, useshrplib=true\, libperl=libperl\.so
>    gnulibc\_version='2\.2\.5'
>  Dynamic Linking&#8203;:
>    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-rdynamic \-Wl\,\-rpath\,/usr/lib/perl5/5\.8\.0/i586\-linux\-thread\-multi/CORE'
>    cccdlflags='\-fPIC'\, lddlflags='\-shared'
>
>Locally applied patches&#8203;:
>    
>
>\-\-\-
>@&#8203;INC for perl v5\.8\.0&#8203;:
>    /usr/lib/perl5/5\.8\.0/i586\-linux\-thread\-multi
>    /usr/lib/perl5/5\.8\.0
>    /usr/lib/perl5/site\_perl/5\.8\.0/i586\-linux\-thread\-multi
>    /usr/lib/perl5/site\_perl/5\.8\.0
>    /usr/lib/perl5/site\_perl
>    \.
>
>\-\-\-
>Environment for perl v5\.8\.0&#8203;:
>    HOME=/suse/mls
>    LANG \(unset\)
>    LANGUAGE \(unset\)
>    LC\_COLLATE=POSIX
>    LC\_CTYPE=de\_DE@&#8203;euro
>    LD\_LIBRARY\_PATH \(unset\)
>    LOGDIR \(unset\)
>    PATH=/suse/mls/bin&#8203;:/opt/kde3/bin&#8203;:/usr/local/bin&#8203;:/usr/bin&#8203;:/usr/sbin&#8203;:/sbin&#8203;:/usr/X11R6/bin&#8203;:/bin&#8203;:/usr/lib/java/bin&#8203;:/usr/games/bin&#8203;:/usr/games&#8203;:/opt/gnome/bin&#8203;:/opt/kde/bin&#8203;:/usr/openwin/bin&#8203;:/opt/pilotsdk/bin&#8203;:/suse/mls/korn
>    PERL\_BADLANG \(unset\)
>    SHELL=/bin/tcsh
\-\- 
Nick Ing\-Simmons
http&#8203;://www\.ni\-s\.u\-net\.com/

**Comment by p5pRT** [comment]
### From mls@suse.de

> This is a side effect of the "safe signals" support\.
> I "had" to turn off SA\_RESTART so that syscall would resturn and give us 
> a chance to execute the perl level %SIG handler\.

Yes\, actually I have complained about breaking compatibility a few
weeks ago \(it was about an EINTR in an accept\(\) call\, yes\, a bit of an
esoteric case\)\. But the current problem is

1\) probably causing many people headaches\, as it is easily to
&nbsp;   trigger\.

2\) easy to fix\. Just retry the stdio call if fgets\(\) returns EOF /
&nbsp;   fread\(\) returns zero and errno is EINTR\. You should probably
&nbsp;   clear errno before the call "just in case"\.

Cheers\,
&nbsp;  Michael\.

\-\- 
Michael Schroeder                                   mls@&#8203;suse\.de
main\(\_\)\{while\(\_=~getchar\(\)\)putchar\(~\_\-1/\(~\(\_|32\)/13\*2\-11\)\*13\);\}

**Comment by p5pRT** [comment]
### From nick.ing-simmons@elixent.com

Michael Schroeder \<mls@&#8203;suse\.de> writes&#8203;:
>> This is a side effect of the "safe signals" support\.
>> I "had" to turn off SA\_RESTART so that syscall would resturn and give us 
>> a chance to execute the perl level %SIG handler\.
>
>Yes\, actually I have complained about breaking compatibility a few
>weeks ago \(it was about an EINTR in an accept\(\) call\, yes\, a bit of an
>esoteric case\)\. 

Several people have complained \- but as the perl \< 5\.8 use of signals
is proven to cause core\-dumps if you use %SIG it was decided that the 
new scheme was better\. Will be better still when we get the bugs out ;\-\)

>But the current problem is
>
>1\) probably causing many people headaches\, as it is easily to
>   trigger\.
>
>2\) easy to fix\. Just retry the stdio call if fgets\(\) returns EOF /
>   fread\(\) returns zero and errno is EINTR\. You should probably
>   clear errno before the call "just in case"\.

Agreed in principle \- do you have a verified patch for perlio\.c to hand?

\-\- 
Nick Ing\-Simmons
http&#8203;://www\.ni\-s\.u\-net\.com/

**Comment by p5pRT** [comment]
### From mls@suse.de

> Agreed in principle \- do you have a verified patch for perlio\.c to hand?

The one attached seems to do the trick\. I didn't patch the
PerlIOStdio\_close\(\) function\, though\, because the linux manpage
contains&#8203;:

RETURN VALUE
&nbsp;       Upon successful completion 0 is returned\.  Otherwise\,  EOF
&nbsp;       is  returned and the global variable errno is set to indi­
&nbsp;       cate  the  error\.   In  either  case  any  further  access
&nbsp;       \(including another call to fclose\(\)\) to the stream results
&nbsp;       in undefined behaviour\.

Hmm\, I don't like that "undefined behaviour" part\.

Cheers\,
&nbsp;  Michael\.

\-\- 
Michael Schroeder                                   mls@&#8203;suse\.de
main\(\_\)\{while\(\_=~getchar\(\)\)putchar\(~\_\-1/\(~\(\_|32\)/13\*2\-11\)\*13\);\}

**Comment by p5pRT** [comment]
### From mls@suse.de

<details><summary>perlio.diff</summary>

```diff
--- perlio.c.orig	2003-01-28 13:13:10.000000000 +0000
+++ perlio.c	2003-01-28 16:12:46.000000000 +0000
@@ -2708,20 +2708,25 @@
 {
     FILE *s = PerlIOSelf(f, PerlIOStdio)->stdio;
     SSize_t got = 0;
-    if (count == 1) {
-	STDCHAR *buf = (STDCHAR *) vbuf;
-	/*
-	 * Perl is expecting PerlIO_getc() to fill the buffer Linux's
-	 * stdio does not do that for fread()
-	 */
-	int ch = PerlSIO_fgetc(s);
-	if (ch != EOF) {
-	    *buf = ch;
-	    got = 1;
+    for (;;) {
+	if (count == 1) {
+	    STDCHAR *buf = (STDCHAR *) vbuf;
+	    /*
+	     * Perl is expecting PerlIO_getc() to fill the buffer Linux's
+	     * stdio does not do that for fread()
+	     */
+	    int ch = PerlSIO_fgetc(s);
+	    if (ch != EOF) {
+		*buf = ch;
+		got = 1;
+	    }
 	}
+	else
+	    got = PerlSIO_fread(vbuf, 1, count, s);
+	if (got || errno != EINTR)
+	    break;
+	errno = 0;	/* just in case */
     }
-    else
-	got = PerlSIO_fread(vbuf, 1, count, s);
     return got;
 }
 
@@ -2786,8 +2791,15 @@
 SSize_t
 PerlIOStdio_write(pTHX_ PerlIO *f, const void *vbuf, Size_t count)
 {
-    return PerlSIO_fwrite(vbuf, 1, count,
-			  PerlIOSelf(f, PerlIOStdio)->stdio);
+    SSize_t got;
+    for (;;) {
+	got = PerlSIO_fwrite(vbuf, 1, count,
+			      PerlIOSelf(f, PerlIOStdio)->stdio);
+	if (got || errno != EINTR)
+	    break;
+	errno = 0;	/* just in case */
+    }
+    return got;
 }
 
 IV

```
</details>

**Comment by p5pRT** [comment]
### From perl5-porters@ton.iguana.be

In article \<20030128130721\.2002\.3@&#8203;bactrian\.elixent\.com>\,
&nbsp;    Nick Ing\-Simmons \<nick\.ing\-simmons@&#8203;elixent\.com> writes&#8203;:
> Michael Schroeder \<mls@&#8203;suse\.de> writes&#8203;:
> 
>>But the current problem is
>>
>>1\) probably causing many people headaches\, as it is easily to
>>   trigger\.
>>
>>2\) easy to fix\. Just retry the stdio call if fgets\(\) returns EOF /
>>   fread\(\) returns zero and errno is EINTR\. You should probably
>>   clear errno before the call "just in case"\.
> 
> Agreed in principle \- do you have a verified patch for perlio\.c to hand?
> 

Sounds dangerous\. errno has no particular guaranteed values after
stdio function failures\. e\.g\. I can imagine it dropping semaphores in
threaded libc's that can sometimes change errno\.

**Comment by p5pRT** [comment]
### From mls@suse.de

> Sounds dangerous\. errno has no particular guaranteed values after
> stdio function failures\. e\.g\. I can imagine it dropping semaphores in
> threaded libc's that can sometimes change errno\.

No it's not dangerous\. The operation will only be retried if it
failed AND errno is EINTR\. So maybe there are some rare cases where
the fix doesn't work \(a failing semaphore operation changes errno from
EINTR to something else\)\, but the patch makes things only better\,
not worse\.

Cheers\,
&nbsp;  Michael\.

\-\- 
Michael Schroeder                                   mls@&#8203;suse\.de
main\(\_\)\{while\(\_=~getchar\(\)\)putchar\(~\_\-1/\(~\(\_|32\)/13\*2\-11\)\*13\);\}

**Comment by p5pRT** [comment]
### From nick@ing-simmons.net

Michael Schroeder \<mls@&#8203;suse\.de> writes&#8203;:
>> Agreed in principle \- do you have a verified patch for perlio\.c to hand?
>
>The one attached seems to do the trick\. I didn't patch the
>PerlIOStdio\_close\(\) function\, though\, because the linux manpage
>contains&#8203;:
>

I have applied the patch to //depot/perlio/perlio\.c@&#8203;18611

I added calls to PERL\_ASYNC\_CHECK\(\) in the loops so that posted %SIG 
callbacks are dispatched before doing re\-try \(which matches 
what :unix and :perlio do\)\. 


\-\- 
Nick Ing\-Simmons
http&#8203;://www\.ni\-s\.u\-net\.com/

**Comment by p5pRT** [comment]
### From @jhi

The patch got applied\, so I'm marking the problem ticket as resolved\.

**Comment by p5pRT** [comment]
@jhi - Status changed from 'new' to 'resolved'
